File: parallelSortGo/src/main/Sort.go
package main

import "sync"

func SortWrapper(arr []int, start int, end int, wg *sync.WaitGroup) {
	defer wg.Done()
	Sort(arr, start, end)
}

func Sort(arr []int, start int, end int) {

	if start >= end {
		return
	}

	pt := start - 1
	for i := start; i < end; i++ {
		if arr[i] < arr[end] {
			pt++
			arr[pt], arr[i] = arr[i], arr[pt]
		}
	}
	pt++
	arr[pt], arr[end] = arr[end], arr[pt]

	Sort(arr, start, pt-1)
	Sort(arr, pt+1, end)
}

File: parallelSortJava/src/main/Sort.java
package main;

public class Sort implements Runnable {

    int[] arr;
    int start;
    int end;


    public Sort(int[] arr, int start, int end) {
        this.arr = arr;
        this.start = start;
        this.end = end;
    }


    public void sort(int l, int r) {
        if (l >= r) {
            return;
        }

        int pt = l - 1;
        for (int i = l; i < r; i++) {
            if (arr[i] < arr[r]) {
                pt++;
                swap(pt, i);
            }
        }
        pt++;
        swap(pt, r);

        sort(l, pt - 1);
        sort(pt + 1, r);

    }

    public void swap(int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    @Override
    public void run() {
        sort(start, end);
    }
}
